netty/src/main/java/io/grpc/netty/GrpcHttp2HeadersUtils.java

```diff
@@ -101,17 +101,34 @@ protected GrpcHttp2InboundHeaders newHeaders() {
       values = new AsciiString[numHeadersGuess];
     }
 
+    @SuppressWarnings("BetaApi") // BaseEncoding is stable in Guava 20.0
     protected Http2Headers add(AsciiString name, AsciiString value) {
-      if (namesAndValuesIdx == namesAndValues.length) {
-        expandHeadersAndValues();
-      }
+
       byte[] nameBytes = bytes(name);
-      byte[] valueBytes = toBinaryValue(name, value);
-      values[namesAndValuesIdx / 2] = value;
-      namesAndValues[namesAndValuesIdx] = nameBytes;
-      namesAndValuesIdx++;
-      namesAndValues[namesAndValuesIdx] = valueBytes;
-      namesAndValuesIdx++;
+      byte[] valueBytes;
+      int startPos = 0;
+      int endPos;
+      int indexOfComma = AsciiString.INDEX_NOT_FOUND;
+      do {
+        AsciiString curVal = value;
+        if (!name.endsWith(binaryHeaderSuffix)) {
+          valueBytes = bytes(value);
+        } else {
+          indexOfComma = value.indexOf(',', startPos);
+          endPos = indexOfComma == AsciiString.INDEX_NOT_FOUND ? value.length() : indexOfComma;
+          curVal = value.subSequence(startPos, endPos, false);
+          valueBytes = BaseEncoding.base64().decode(curVal);
+          startPos = indexOfComma + 1;
+        }
+        if (namesAndValuesIdx == namesAndValues.length) {
```

Comment:
It's hard to see the flow now, especially for ascii values. Could the actual adding to the array be moved into a private method, like `add0(byte[] name, byte[] value, AsciiString valueStr)`? That would let you use the loop only for binary values.